File: frontend/src/types/report.ts
export type ReportStatus =
  | "UPLOADED"
  | "PROCESSING"
  | "EXTRACTION_FAILED"
  | "READY_TO_PUBLISH"
  | "PUBLISHED"
  | "ARCHIVED";

export interface ExtractedField {
  value: string | number | null;
  confidence: number;
  type: "text" | "number" | "currency" | "date";
  original?: string | number | null;
  edited?: boolean;
}

export interface ContentJson {
  extraction_version: number;
  provider: string;
  model: string;
  anchor_fields: Record<string, ExtractedField>;
  additional_fields: Record<string, ExtractedField>;
  raw_text: string;
  extracted_at: string;
  page_count: number;
  usage: { input_tokens: number; output_tokens: number };
}

export interface Report {
  id: string;
  request_id: string | null;
  vendor_id: string;
  report_category: "VALUATION" | "LEGAL";
  status: ReportStatus;
  property_address: string | null;
  macro_location: string | null;
  city: string | null;
  pin_code: string | null;
  property_type: string | null;
  plot_extent_sqft: string | null;
  built_up_sqft: string | null;
  valuation_amount: string | null;
  loan_applicant_name: string | null;
  report_date: string | null;
  expiry_date: string | null;
  uploaded_file_path: string | null;
  content_json: ContentJson | null;
  listing_approved: boolean;
  is_active: boolean;
  latitude?: string | null;
  longitude?: string | null;
}

export interface ReportRevision {
  revision_number: number;
  comments: string | null;
  created_at: string;
}

File: backend/app/schemas/report.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel


class ReportCreate(BaseModel):
    vendor_id: UUID
    report_category: str
    property_address: str | None = None
    macro_location: str | None = None
    city: str | None = None
    pin_code: str | None = None
    property_type: str | None = None
    plot_extent_sqft: Decimal | None = None
    built_up_sqft: Decimal | None = None
    valuation_amount: Decimal | None = None
    loan_applicant_name: str | None = None
    report_date: date | None = None
    uploaded_file_path: str | None = None
    latitude: Decimal | None = None
    longitude: Decimal | None = None


class ReportResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    vendor_id: UUID
    report_category: str
    status: str
    property_address: str | None = None
    macro_location: str | None = None
    city: str | None = None
    pin_code: str | None = None
    property_type: str | None = None
    plot_extent_sqft: Decimal | None = None
    built_up_sqft: Decimal | None = None
    valuation_amount: Decimal | None = None
    loan_applicant_name: str | None = None
    report_date: date | None = None
    expiry_date: date | None = None
    uploaded_file_path: str | None = None
    latitude: Decimal | None = None
    longitude: Decimal | None = None
    content_json: dict | None = None
    listing_approved: bool
    is_active: bool


class ReportBrief(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    report_category: str
    city: str | None = None
    macro_location: str | None = None
    property_type: str | None = None
    status: str
    report_date: date | None = None


class ReportUploadMeta(BaseModel):
    """Optional metadata submitted with report upload."""
    valuation_amount: Decimal | None = None
    report_date: date | None = None


class RevisionSummary(BaseModel):
    model_config = {"from_attributes": True}

    revision_number: int
    comments: str | None = None
    created_at: datetime | None = None


class ReportDetail(ReportResponse):
    """Extended response with revision history."""
    revisions: list[RevisionSummary] = []


class ExtractedFieldUpdate(BaseModel):
    value: str | int | float | None = None
    confidence: float | None = None
    type: str = "text"
    original: str | int | float | None = None
    edited: bool = False


class ExtractedDataUpdate(BaseModel):
    """Payload for updating extracted report data."""
    anchor_fields: dict[str, ExtractedFieldUpdate] = {}
    additional_fields: dict[str, ExtractedFieldUpdate] = {}
